File: examples/typescript/case08.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import {
  WeakRecoveryProjection,
  StrongRecoveryProjection,
  persistStateWithRevision,
  rebuildBeforeReady,
  type RecoveryStore,
} from "./case08.ts";

test("FS-C08 loses only the volatile guard, then recovers it coherently", () => {
  const weakStore: RecoveryStore = { record: { state: "", revision: 0 } };
  const before = new WeakRecoveryProjection(weakStore);
  assert.equal(before.apply({ state: "SHIPPED", revision: 44 }), true);
  assert.equal(before.apply({ state: "PROCESSING", revision: 42 }), false);
  assert.equal(weakStore.record.state, "SHIPPED");
  const after = new WeakRecoveryProjection(weakStore);
  assert.equal(after.apply({ state: "PROCESSING", revision: 42 }), true);
  assert.equal(weakStore.record.state, "PROCESSING");

  const strongStore: RecoveryStore = { record: { state: "", revision: 0 } };
  persistStateWithRevision(strongStore, { state: "SHIPPED", revision: 44 });
  const recovered = StrongRecoveryProjection.recover(strongStore);
  assert.equal(recovered.apply({ state: "PROCESSING", revision: 42 }), false);
  assert.deepEqual(strongStore.record, { state: "SHIPPED", revision: 44 });
  assert.equal(recovered.apply({ state: "DELIVERED", revision: 45 }), true);
  assert.deepEqual(strongStore.record, { state: "DELIVERED", revision: 45 });

  const rebuilt = rebuildBeforeReady([
    { state: "PROCESSING", revision: 42 },
    { state: "SHIPPED", revision: 44 },
  ]);
  assert.equal(rebuilt.apply({ state: "PROCESSING", revision: 42 }), false);
});

File: examples/python/case08_test.py
import unittest

from case08 import (
    RecoveryRecord,
    RecoveryStore,
    WeakRecoveryProjection,
    StrongRecoveryProjection,
    persist_state_with_revision,
    rebuild_before_ready,
)


class Case08RecoveryTest(unittest.TestCase):
    def test_recovery_restores_revision_evidence(self) -> None:
        weak_store = RecoveryStore(RecoveryRecord("", 0))
        before = WeakRecoveryProjection(weak_store)
        self.assertTrue(before.apply(RecoveryRecord("SHIPPED", 44)))
        self.assertFalse(before.apply(RecoveryRecord("PROCESSING", 42)))
        self.assertEqual(weak_store.record.state, "SHIPPED")
        after = WeakRecoveryProjection(weak_store)
        self.assertTrue(after.apply(RecoveryRecord("PROCESSING", 42)))
        self.assertEqual(weak_store.record.state, "PROCESSING")

        strong_store = RecoveryStore(RecoveryRecord("", 0))
        persist_state_with_revision(strong_store, RecoveryRecord("SHIPPED", 44))
        recovered = StrongRecoveryProjection.recover(strong_store)
        self.assertTrue(recovered.ready)
        self.assertFalse(recovered.apply(RecoveryRecord("PROCESSING", 42)))
        self.assertEqual(strong_store.record, RecoveryRecord("SHIPPED", 44))
        self.assertTrue(recovered.apply(RecoveryRecord("DELIVERED", 45)))
        self.assertEqual(strong_store.record, RecoveryRecord("DELIVERED", 45))

        rebuilt = rebuild_before_ready([
            RecoveryRecord("PROCESSING", 42), RecoveryRecord("SHIPPED", 44)
        ])
        self.assertFalse(rebuilt.apply(RecoveryRecord("PROCESSING", 42)))


if __name__ == "__main__":
    unittest.main()
